crates/ruff_server/src/session/settings.rs

```diff
@@ -179,9 +210,68 @@ impl ResolvedClientSettings {
                 },
                 true,
             ),
+            editor_settings: ResolvedEditorSettings {
+                lint_preview: Self::resolve_optional(all_settings, |settings| {
+                    settings.lint.as_ref()?.preview
+                }),
+                format_preview: Self::resolve_optional(all_settings, |settings| {
+                    settings.format.as_ref()?.preview
+                }),
+                select: Self::resolve_optional(all_settings, |settings| {
+                    settings
+                        .lint
+                        .as_ref()?
+                        .select
+                        .as_ref()?
+                        .iter()
+                        .map(|rule| RuleSelector::from_str(rule).ok())
+                        .collect()
+                }),
+                extend_select: Self::resolve_optional(all_settings, |settings| {
+                    settings
+                        .lint
+                        .as_ref()?
+                        .extend_select
+                        .as_ref()?
+                        .iter()
+                        .map(|rule| RuleSelector::from_str(rule).ok())
+                        .collect()
+                }),
+                ignore: Self::resolve_optional(all_settings, |settings| {
+                    settings
+                        .lint
+                        .as_ref()?
+                        .ignore
+                        .as_ref()?
+                        .iter()
+                        .map(|rule| RuleSelector::from_str(rule).ok())
+                        .collect()
+                }),
+                exclude: Self::resolve_optional(all_settings, |settings| {
+                    Some(
+                        settings
+                            .exclude
+                            .as_ref()?
+                            .iter()
+                            .map(|path| PathBuf::from(OsString::from(path)))
+                            .collect(),
+                    )
+                }),
+                line_length: Self::resolve_optional(all_settings, |settings| settings.line_length),
+            },
         }
     }
 
+    /// Attempts to resolve a setting using a list of available client settings as sources.
+    /// Client settings that come earlier in the list take priority. This function is for fields
+    /// that do not have a default value and should be left unset.
+    fn resolve_optional<T>(
```

Comment:
I should probably explain why I'm not just removing `ResolvedClientSettings::resolve_or` entirely and using `resolve_optional(...).unwrap_or(<default>)` in its place. My reasoning is that I don't want to move the logic that determines default values out of a dedicated function, because it opens up the possibility that someone might use `resolve_optional(...).unwrap_or_default()` or some other incorrect method of resolving a default value. We should always make the default value explicit when resolving a setting.

Reply:
Maybe the documentation of `resolve_or` could point to `resolve_optional` and similarly you could add `resolve_or` in `resolve_optional`'s documentation. The wording would be:
```rs
// Use [`ResolvedClientSettings::resolve_optional`] for ...
```